File: 백준/Silver/2839. 설탕 배달/설탕 배달.py
import sys
input = sys.stdin.readline
N = int(input())
possible = [-1] * (N+1)
possible[N] = 0

for i in range(N, -1, -1):
  if possible[i] != -1:
    if 0 <= i-5:
      if possible[i-5] == -1:
        possible[i-5] = possible[i] + 1
      elif possible[i-5] > possible[i] + 1:
        possible[i-5] = possible[i] + 1
    if 0 <= i-3:
      if possible[i-3] == -1:
        possible[i-3] = possible[i] + 1
      elif possible[i-3] > possible[i] + 1:
        possible[i-3] = possible[i] + 1   
print(possible[0])

File: 백준/Silver/2839. 설탕 배달/설탕 배달.java
import java.io.IOException;
import java.util.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] possible = new int[N+1];
        Arrays.fill(possible, -1);
        possible[N] = 0;
        for(int i=N; i>=0; i--) {
            if(possible[i] != -1) {
                if(i-5 >= 0){
                    if(possible[i-5] == -1){
                    possible[i-5] = possible[i] + 1;
                    }
                    else{
                        possible[i-5] = Math.min(possible[i-5], possible[i] + 1);
                    }
                }
                if(i-3 >= 0){
                    if(possible[i-3] == -1){
                        possible[i-3] = possible[i] + 1;
                        }
                    else{
                        possible[i-3] = Math.min(possible[i-3], possible[i] + 1);
                    }
                }
            }
            //System.out.println(Arrays.toString(possible));
        }
        System.out.println(possible[0]);
    }
}